lightning/src/routing/router.rs

```diff
@@ -222,67 +329,121 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 		}
 	}
 
+	// We don't want multiple paths (as per MPP) share capacity of the same channels.
+	// This map allows paths to be aware of the channel use by other paths in the same call.
+	// This would help to make a better path finding decisions and not "overbook" channels.
+	// It is currently unaware of the directions. But if we moved 1 BTC in one direction and
+	// 1 BTC in the opposite direction, should they cancel out? Probably not, because
+	// in the worst-case order of HTLC forwarding, a channel can be overspent.
+	let mut known_channels_amounts_available_msat = HashMap::new();
+
+	// Keeping track of how much value we need to collect across channels at this iteration,
+	// so that we know which channels are definitely not worth considering: e.g., if their
+	// available capacity is only a small fraction of what we need.
+	// Also know when to stop looking for new paths.
+	let mut found_amount_msat = 0;
+
 	macro_rules! add_entry {
 		// Adds entry which goes from $src_node_id to $dest_node_id
 		// over the channel with id $chan_id with fees described in
 		// $directional_info.
-		( $chan_id: expr, $src_node_id: expr, $dest_node_id: expr, $directional_info: expr, $chan_features: expr, $starting_fee_msat: expr ) => {
+		( $chan_id: expr, $src_node_id: expr, $dest_node_id: expr, $directional_info: expr, $capacity_sats: expr, $chan_features: expr, $starting_fee_msat: expr ) => {
+			let amount_available_msat = known_channels_amounts_available_msat.entry($chan_id.clone()).or_insert_with(|| {
+				let mut channel_amount_available_msat = None;
+				if let Some(capacity_sats) = $capacity_sats {
+					channel_amount_available_msat = Some(capacity_sats * 1000);
+				}
+
+				if let Some(htlc_maximum_msat) = $directional_info.htlc_maximum_msat {
+					if let Some(available_msat) = channel_amount_available_msat {
+						channel_amount_available_msat = Some(cmp::min(available_msat, htlc_maximum_msat));
+					} else {
+						channel_amount_available_msat = Some(htlc_maximum_msat);
+					}
+				}
+
+				match channel_amount_available_msat {
+					Some(available_msat) => available_msat,
+					// We assume channels with unknown balance have a capacity of 0.0001 BTC (or 10_000 sats).
+					None => 10_000 * 1000
+				}
+			});
+
+			// Skip this channel if its available capacity is very low comparing to what we need at the moment.
+			let has_sufficient_amount = *amount_available_msat * 20 >= recommended_available_msat - found_amount_msat;
```

Comment:
>I think it would simplify to drop the 20 and increase the multiplicand of recommended_available_msat, same effect ?

Not increase, reduce. Improving the comments and code structure around here.